Add EIP-5656: Memory Copying Instruction (#5656)
ethereum/EIPs: EIPS/eip-5656.md

---
eip: 5656

> title: Memory copying instruction

Reviewer: ```suggestion
title: Optcode Instruction for Memory Copy
```

Reviewer 2: I think this is unnecessary + it's "opcode" not "opt".

Reviewer: Makes sense @lightclient.
I withdraw my suggested revision.

description: An efficient EVM instruction for copying memory areas
author: Alex Beregszaszi (@axic), Paul Dworzanski (@poemm), Jared Wasinger (@jwasinger), Casey Detrio (@cdetrio), Pawel Bylica (@chfast), Charles Cooper (@charles-cooper)
discussions-to: https://ethereum-magicians.org/t/eip-5656-mcopy-instruction/10890
status: Draft
type: Standards Track
category: Core
created: 2021-02-01
---

## Abstract

> Provide an efficient EVM instruction for copying memory areas.

Reviewer 3: This should be (at a minimum) a full descriptive sentence describing a high-level overview 
of the solution that this EIP provides:
```suggestion
This EIP defines a new EVM opcode, `MCOPY`, for efficiently copying continuous sections of memory.
```

Reviewer: Yep, good copywrite, this's much more readable. @Pandapip1

Author: This defines a new instruction. Not opcode which is just a number.

Reviewer 3: Opcodes can be referred to either by number or by name. Feel free to correct me if I am wrong though.

Reviewer 4: I also prefer the current status quo of the distinction between instruction and opcode.

## Motivation

> Memory copying is a basic operation, yet implementing it on the EVM comes with overhead.

Reviewer 3: ```suggestion
Memory copying is a frequent operation, yet implementing it on the EVM comes with significant overhead.
```
Basic != Frequently Used
Significant: It's going to come with overhead, the point is that right now it's too much overhead.

- 27 gas using this EIP

According to an analysis of blocks 10537502 to 10538702, roughly 10.5% of memory copies would have had improved performance with the
availability of an `MCOPY` instruction.

Memory copying is used by languages like Solidity and Vyper, where we expect this improvement to provide efficient means of building
data structures, including efficient sliced access and copies of memory objects. Having a dedicated `MCOPY` instruction would also add
forward protection against future gas cost changes to `CALL` instructions in general.

Having a special `MCOPY` instruction makes the job of static analyzers and optimizers easier, since the effects of a `CALL` in general
have to be fenced, whereas an `MCOPY` instruction would be known to only have memory effects. Even if special cases are added
for precompiles, a future hard fork could change `CALL` effects, and so any analysis of code using the identity precompile would only
be valid for a certain range of blocks.

Finally, we expect memory copying to be immensely useful for various computationally heavy operations, such as EVM384,
where it is identified as a significant overhead.

## Specification

> The instruction `MCOPY` is introduced at `0x5c`.

Reviewer: Should we use a parameter placeholder for the instruction instead of `0x5c` and then list this instruction when a fork is determined? Other EIP might be introducing new optcode in the meanwhile.

Reviewer 4: A number must be decided in order to implement it in clients and to generate tests. This has been the best practice. In case it gets CFId/scheduled/etc. the numbers can be changed.

> The instruction `MCOPY` is introduced at `0x5c`.

Reviewer 4: > RFC 2119 defines some pretty important terms. It should be used.

Good point, however EIP-1 states it is optional, and we actually don't use these term in the spec.

### Input stack

| Stack | Value |
|-------|-------|
| top - 0 | `dst` |
| top - 1 | `src` |
| top - 2 | `length` |

> This ordering matches the other copying instructions, i.e. `CALLDATACOPY`, `RETURNDATACOPY`.

Reviewer 3: Should be moved to the rationale section.

### Output stack

This instructions returns no stack items.

### Semantics

> It copies `length` bytes from the offset pointed at `src` to the offset pointed at `dst` in memory.

Reviewer 4: I think this is really splitting hairs, but happy to change it.

Copying takes place as if an intermediate buffer were used, allowing the destination and source to overlap.

If `length > 0` and (`src + length` or `dst + length`) is beyond the current memory length, the memory is extended with respective gas cost applied.

The gas cost of this instruction mirrors that of other `Wcopy` instructions and is `Gverylow + Gcopy * ceil(length / 32)`.

## Rationale

Production implementation of exact-word memory copying and partial-word memory copying can be found in the Solidity, Vyper and Fe compilers.

With [EIP-2929](./eip-2929.md) the call overhead using the identity precompile was reduced from 700 to 100 gas.
This is still prohibitive for making the precompile a reasonable alternative again.

## Backwards Compatibility

> This EIP introduces a new instruction which did not exists previously. Already deployed contracts using this instruction could change their behaviour after this EIP.

Reviewer 5: Since this opcode would be added after EOF is introduced there will be no conflict. with pre-EOF code.

## Test Cases

TBA

## Security Considerations

> TBA

Reviewer 3: ```suggestion
Needs discussion.
```
My personal preference. Feel free to ignore it.

Reviewer 5: I think EIP-1 demands a security discussion, if only to explain why there are no impacts.

Reviewer 3: This is correct, but the restrictions are more relaxed for drafts.

---
eip: 5656
title: Memory copying instruction
description: An efficient EVM instruction for copying memory areas
author: Alex Beregszaszi (@axic), Paul Dworzanski (@poemm), Jared Wasinger (@jwasinger), Casey Detrio (@cdetrio), Pawel Bylica (@chfast), Charles Cooper (@charles-cooper)

> discussions-to: <tba>

Reviewer 3: Needs a URL

status: Draft
type: Standards Track
category: Core
created: 2021-02-01
---

## Abstract

Provide an efficient EVM instruction for copying memory areas.

## Motivation

Memory copying is a basic operation, yet implementing it on the EVM comes with overhead.

This was recognised and alleviated early on with the introduction of the "identity" precompile, which accomplishes
memory copying by the use of `CALL`'s input and output memory offsets. Its cost is `15 + 3 * (length / 32)` gas, plus
the call overhead. The identity precompile was rendered ineffective by the raise of the cost of `CALL` to 700, but subsequently

> the reduction by [EIP-2929](./eip-2929) made it slightly more economical.

Reviewer 3: ```suggestion
the reduction by [EIP-2929](./eip-2929.md) made it slightly more economical.
```

for precompiles, a future hard fork could change `CALL` effects, and so any analysis of code using the identity precompile would only
be valid for a certain range of blocks.

Finally, we expect memory copying to be immensely useful for various computationally heavy operations, such as EVM384,
where it is identified as a significant overhead.

## Specification

Starting `BLOCK_NUMBER >= HF`, a new instruction, `MCOPY` (`0x5c`) is introduced.

It takes three arguments off the stack (1. represents the top of the stack):
1. `dst`
2. `src`
3. `length`

It copies `length` bytes from the offset pointed at `src` to the offset pointed at `dst` in memory.
Copying takes place as if an intermediate buffer were used, allowing the destination and source to overlap.

If `length > 0` and (`src + length` or `dst + length`) is beyond the current memory length, the memory is extended with respective gas cost applied.

The gas cost of this instruction mirrors that of other `Wcopy` instructions and is `Gverylow + Gcopy * ceil(length / 32)`.

## Rationale

Production implementation of exact-word memory copying and partial-word memory copying can be found in the Solidity, Vyper and Fe compilers.

> With [EIP-2929](https://eips.ethereum.org/EIPS/eip-2929) the call overhead using the identity precompile was reduced from 700 to 100 gas.

Reviewer 3: ```suggestion
With [EIP-2929](./eip-2929.md) the call overhead using the identity precompile was reduced from 700 to 100 gas.
```

This is still prohibitive for making the precompile a reasonable alternative again.

## Backwards Compatibility

This EIP introduces a new instruction which did not exists previously. Already deployed contracts using this instruction could change their behaviour after this EIP.

> ## Test Cases
> 
> TBA

Reviewer 3: ```suggestion
```

Reviewer 4: If possible would like to keep as it is nice to include some core test cases and this is a reminder for us.

Reviewer 3: I would be okay with this:
```suggestion
<!-- TODO: Test Cases -->
```

Reviewer 4: Comments are not rendered, so they are not very useful.

Reviewer 3: That's the point. They aren't rendered, so they aren't shown unless you're editing the source file.

## Security Considerations

TBA

> ## Copyright
> 
> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer 3: ```suggestion
## Copyright

Copyright and related rights waived via [CC0](../LICENSE.md).
```
